src/commands/heroku.rs

```diff
@@ -0,0 +1,66 @@
+use heroku_rs::client::{Executor, Heroku};
+
+use serde_json::Value;
+
+use serenity::framework::standard::{macros::command, Args, CommandResult};
+use serenity::model::prelude::*;
+use serenity::prelude::*;
+
+use crate::config::Config;
+
+#[derive(Debug)]
+struct HerokuApp {
+    id: String,
+    name: String,
+}
+
+#[command]
+pub fn get_apps(ctx: &mut Context, msg: &Message, _args: Args) -> CommandResult {
+    let response = heroku_client().get().apps().execute::<Value>();
+    let mut processed_app_list: Vec<HerokuApp> = Vec::new();
+
+    match response {
+        Ok((_headers, _status, json)) => {
+            if let Some(json) = json {
+                let array = json.as_array().unwrap();
+                processed_app_list.append(&mut app_list(array))
+            }
+        }
+
+        Err(e) => println!("Err {}", e),
+    }
+
+    msg.reply(ctx, app_response(processed_app_list))?;
+
+    Ok(())
+}
+
+fn heroku_client() -> heroku_rs::client::Heroku {
+    let heroku_api_key = Config::default().heroku_api_key;
+    Heroku::new(heroku_api_key).unwrap()
```

Comment:
The `Heroku` struct embeds a HTTP client, so if we reuse it across requests instead of creating a new one each command we could rely on keep-alive connections to get a performance boost.

Reply:
I experimented with creating a heroku client in the run() method in lib.rs. I wanted to pass the client to each of the commands. Unfortunately, the only way to do this is to write to the data attribute on the Serenity Client. In order to pass an instance of a struct to the data attribute, it needs to implement the TypeMapKey trait (which is defined in Serenity). The Heroku struct is defined in the Heroku-rs crate...so unfortunately I cannot implement the TypeMapKey trait on the Heroku struct as it is right now. I'm thinking about leaving it as creating a Heroku client for each command for now, but if we are having performance issues we can certainly revisit this.

Reply:
I don't think implementing `TypeMapKey` for the `Heroku` struct is needed, as the key doesn't need to be the same type as the value. You could do something like this:

```rust
pub(crate) struct HerokuClient;

impl TypeMapKey for HerokuClient {
    type Value = heroku_rs::client::Client;
}
```

...and then use the `HerokuClient` key to retrieve the proper client.

Reply:
I wouldn't be particularly concerned with performance for the bot – it will probably never get more than one request per minute or so. However, creating the Heroku client creates an event loop, a thread pool and a reactor per thread. Doing that for each request just _feels_ a bit too much. :)

(I'm also a bit surprised that the client creates the event loop internally. If it needs one, I'd expect it to accept the event loop as a parameter.)

Reply:
I've been attempting this with:
```rust
use heroku_rs::client::{Heroku};

pub(crate) struct HerokuClient;

impl TypeMapKey for HerokuClient {
    type Value = heroku_rs::client::Heroku;
}
```

and 

```rust
 data.insert::<HerokuClient>(heroku_client);
```

However, this is returning a number of errors:

```bash
error[E0277]: `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be sent between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be sent between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be sent between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be sent between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Send` is not implemented for `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be shared between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>` cannot be shared between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Sync` is not implemented for `std::rc::Rc<std::cell::RefCell<tokio_core::reactor::Core>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error[E0277]: `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be shared between threads safely
  --> src/lib.rs:47:14
   |
47 |         data.insert::<HerokuClient>(heroku_client);
   |              ^^^^^^ `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>` cannot be shared between threads safely
   |
   = help: within `heroku_rs::client::Heroku`, the trait `std::marker::Sync` is not implemented for `std::rc::Rc<hyper::client::Client<hyper_tls::client::HttpsConnector<hyper::client::connect::http::HttpConnector>>>`
   = note: required because it appears within the type `heroku_rs::client::Heroku`
   = note: required because of the requirements on the impl of `typemap::internals::Implements<(dyn unsafe_any::UnsafeAny + std::marker::Send + std::marker::Sync + 'static)>` for `heroku_rs::client::Heroku`

error: aborting due to 4 previous errors
```

This appears to be because of how the fields are defined for [the Heroku struct in heroku-rs](https://github.com/bensadiku/heroku-rs/blob/d3e97651c840d2cbf5d7892619329bfa0b908340/src/client.rs#L33)

```rust
#[derive(Debug)]
pub struct Heroku {
    token: String,
    core: Rc<RefCell<Core>>,
    client: Rc<Client<HttpsConnector>>,
}
```

Rc<T> and RefCell<T> are, to my knowledge, not shareable across threads.

I will be looking into contributing to heroku-rs, to see if I can make the client safe to share across threads, but it feels a bit out of scope for this pull request.  Once this is merged, I am happy to create an issue to figure out how to share the Heroku client across threads, including contributing to heroku-rs.

Reply:
Issue has been created! https://github.com/rust-lang/crates-io-ops-bot/issues/19

Reply:
Sounds good.